cmd/download.go

```diff
@@ -0,0 +1,250 @@
+		}
+
+		if err != nil {
+			return nil, fmt.Errorf("could not create FHIR server request: %v\n", err)
+		}
+
+		requestStart := time.Now()
+		response, err := client.Do(request)
+		if err != nil {
+			return nil, fmt.Errorf("could not request the FHIR server: %v\n", err)
+		}
+		stats.requestDurations = append(stats.requestDurations, time.Since(requestStart).Seconds())
+
+		procStart := time.Now()
+		responseBody, err := ioutil.ReadAll(response.Body)
+		if err != nil {
+			return nil, fmt.Errorf("could not read FHIR server response: %v\n", err)
+		}
+		response.Body.Close()
+		stats.totalBytesIn += int64(len(responseBody))
+
+		var resourceBundle map[string]interface{}
+		err = json.Unmarshal(responseBody, &resourceBundle)
+		if err != nil {
+			return nil, fmt.Errorf("could not parse FHIR server response: %v\n", err)
+		}
+
+		numResources, err := writeResources(resourceBundle, sink)
+		if err != nil {
+			return nil, fmt.Errorf("could not write resource to output file: %v\n", err)
+		}
+		stats.resourcesPerPage = append(stats.resourcesPerPage, numResources)
+
+		nextPageURL, err = getNextPageURL(resourceBundle)
+		if err != nil {
+			return nil, fmt.Errorf("could not parse the next page link within the FHIR server response: %v\n", err)
+		}
+		stats.processingDurations = append(stats.processingDurations, time.Since(procStart).Seconds())
+	}
+
+	stats.endTime = time.Now()
+	return &stats, nil
+}
+
+// createOutputFileOrDie creates the output file at the given filepath if it does not already exist
+// and returns the file handle.
+// This is a non-destructive operation. Hence, if a file already exists at the given filepath then
+// the command exits with a non-success error code. If any other error case the command exits with
+// a non-success error code as well.
+//
+// Note: The callee has to make sure that the file handle is closed properly.
+func createOutputFileOrDie(filepath string) *os.File {
+	outputFile, err := os.OpenFile(filepath, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0644)
+	if err != nil {
+		if os.IsExist(err) {
+			fmt.Printf("The output file at `%v` does already exist.\n", filepath)
+			os.Exit(3)
+		} else {
+			fmt.Printf("could not open/create the output file at `%s`: %v", filepath, err)
+			os.Exit(4)
+		}
+	}
+	return outputFile
+}
+
+// writeResources writes all resources of the given resourceBundle as-is to the given sink.
+// Returns the number of written resources or an error if the resources can not be written to the sink.
+// An error is only possible if there is an actual issue writing to the file or the given resource
+// bundle is invalid with regards to the FHIR specification.
+func writeResources(resourceBundle map[string]interface{}, sink io.Writer) (resources int, err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			err = fmt.Errorf("encountered a malformed resource bundle")
+		}
+	}()
+
+	entries := resourceBundle["entry"].([]interface{})
```

Comment:
does this fail if there is no entry property? See link below. Should not fail on no entry.

Reply:
This will fail, yes. After checking the FHIR specification again this is specified as 0..*. So indeed, this should not fail rendering the recovering block superfluous.